Database/CleanArchive.py

```diff
@@ -88,12 +96,69 @@ def clean(all=True):
         transaction = {
             '$pull': {'snippets': {'transcript': ''}}
         }
+        if duplicate < 1.0:
+            print('Duplicate found')
+            set_fields['duplicate'] = True
+            duplicates['metadata.Subtitle'] += 1
+        else:
+            set_fields['duplicate'] = False
         if errors:
             set_fields['errors'] = errors
         else:
             transaction['$unset'] = {'errors': 1}
         if set_fields:  # $set cannot be empty.
             transaction['$set'] = set_fields
+
+
+
         db.ArchiveIndex.update_one({'_id': episode['_id']}, transaction)
     print('Updates:', updates)
     print('Failures:', failures)
+    print('Duplicates', duplicates)
+
+
+def checkDuplicate(epi):
+    upperBound = db.ArchiveIndex.find({'identifier': epi['identifier']})[0]['date']
+    dateTimeObj = datetime.datetime.strptime(upperBound, '%Y-%m-%dT%H:%M:%SZ')
+    lowerBound = str(dateTimeObj - dateTimeObj - datetime.timedelta(days=1))
+    fields = db.ArchiveIndex.find({'date': {'$lte': upperBound, '$gte': lowerBound}}, {'snippets': 1})
+    currentEpisode = ''
+    for snip in epi['snippets']:
+        currentEpisode += str(snip['transcript'])
+    for field in fields:
+        compareEpisode = ''
+        cosine = 0
+        try:
+            for snip in field['snippets']:
+                compareEpisode += str(snip['transcript'])
+            xList = word_tokenize(currentEpisode)
+            yList = word_tokenize(compareEpisode)
+            sw = stopwords.words('english')
+            l1 = []
+            l2 = []
+            xSet = {w for w in xList if not w in sw}
+            ySet = {w for w in yList if not w in sw}
+            vector = xSet.union(ySet)
+            for w in vector:
+                if (w in xSet):
+                    l1.append(1)
+                else:
+                    l1.append(0)
+                if (w in ySet):
+                    l2.append(1)
+                else:
+                    l2.append(0)
+                c = 0
+            for i in range(len(vector)):
+                c += l1[i] * l2[i]
+            cosine = c / float((sum(l1) * sum(l2)) ** .5)
+            # print(cosine)
+
+
+        except Exception as e:
+            print('There was an issue comparing two of the transcripts')
+        if cosine > 0.5:
+            print(cosine)
+            return cosine
```

Comment:
Wouldn't this exit early before considering all episodes?

Reply:
Yes, we are only wondering if it is a duplicate. Do we care how many it matches?

Reply:
I think it's useful to list all matches. Could help with debugging or give interesting information about what shows are re-run a lot.
How did you determine the threshold for similarity? I think we need to perform some tests.

Reply:
Most are in the .2-.3 range when not similar so I chose one higher than that threshold